File: lab01/ruby-adapter/adapter_demo.rb
#*----------------------------------------------------------------------------*\
#  Demos of the Adapter implementation
#  @author Norm Evangelista
#*----------------------------------------------------------------------------*/
require_relative 'shape_interface_draw'
require_relative 'shape_interface_draw_acme'

#*----------------------------------------------------------------------------*\
#  Loads all base ShapeInterfaceDraw objects and draws them
#*----------------------------------------------------------------------------*/
def adapter_legacy
   include ShapeInterface

   shapes = []
   shapes << Point.new
   shapes << Line.new
   shapes << Rect.new

   shapes.each do |shape|
      shape.draw
   end
   puts ""
end

#*----------------------------------------------------------------------------*\
#  Loads all base ShapeInterfaceDraw objects and Adapted Acme shapes and draws them
#*----------------------------------------------------------------------------*/
def adapter_problem
   include ShapeInterface

   shapes = []
   shapes << Point.new
   shapes << Line.new
   shapes << Rect.new

   shapes << Polygon.new
   shapes << Torus.new
   shapes << Bezel.new

   shapes.each do |shape|
      shape.draw
   end
   puts ""
end

#*----------------------------------------------------------------------------*\
#  Call examples
#*----------------------------------------------------------------------------*/
adapter_legacy
adapter_problem

File: lab01/adapter/AdapterDemo.cpp
/*---------------------------------------------------------------------------*\
| \file     AdapterDemo.cpp
| \brief    Various demos of the Adapter implementation
| \author   Norm Evangelista
\*---------------------------------------------------------------------------*/
#include "ShapeInterfaceDraw.hpp"
#include "AcmeShapeInterfaceDisplay.h"

#include <memory>
#include <vector>

/*---------------------------------------------------------------------------*\
|   Convenience typedefs
\*---------------------------------------------------------------------------*/
typedef std::unique_ptr<ShapeInterfaceDraw> TShapePtr;
typedef std::vector<TShapePtr>              TShapeVec;

namespace adapter_legacy
{
/*---------------------------------------------------------------------------*\
| \fn       demo
| \brief    Loads all base ShapeInterfaceDraw objects and draws them
\*---------------------------------------------------------------------------*/
void demo()
{
    TShapeVec shapes;
    shapes.push_back( TShapePtr( new Point ) );
    shapes.push_back( TShapePtr( new Line ) );
    shapes.push_back( TShapePtr( new Rect ) );

    for ( auto& it : shapes )                                         
    {
        it->draw();
    }

    std::cout << std::endl;
}
}

namespace adapter_problem
{
/*---------------------------------------------------------------------------*\
| \fn       demo
| \brief    Loads all base ShapeInterfaceDraw objects and draws them
\*---------------------------------------------------------------------------*/
void demo()
{
    TShapeVec shapes;
    shapes.push_back( TShapePtr( new Point ) );
    shapes.push_back( TShapePtr( new Line ) );
    shapes.push_back( TShapePtr( new Rect ) );
    shapes.push_back( TShapePtr( new Polygon ) );
    shapes.push_back( TShapePtr( new Torus ) );
    shapes.push_back( TShapePtr( new Bezel ) );

    for ( auto& it : shapes )                                         
    {
        it->draw();
    }

    std::cout << std::endl;
}
}


/*---------------------------------------------------------------------------*\
| \fn       main
| \brief    Demos legacy and Adapter Pattern extended interfaces
\*---------------------------------------------------------------------------*/
int main()
{
    adapter_legacy::demo();

    adapter_problem::demo();

    return EXIT_SUCCESS;
}

